Provide jpa metamodel for PanacheEntity class

Postremus opened the pull request:
The created meta model for an entity extends PanacheEntity_.
PanacheEntity_ is not created during the quarkus build, and therefore would have to be created by the user, which is tedius.

We now just run the jpamodelgen in the hiberate-orm-panache modul, creating PanacheEntity_ automaticly during the build.

gsmet commented on extensions/panache/hibernate-orm-panache/runtime/pom.xml:
```diff
@@ -39,6 +39,36 @@
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-jpamodelgen</artifactId>
+            <scope>provided</scope>
+            <!-- Exclude these dependencies, since the are banned -->
```
```suggestion
            <!-- Exclude these dependencies, since they are banned -->
```

Postremus replied:
sure, I will do this change this evening.

gsmet reviewed (changes requested):
Thanks.

I'll let @FroMage decide but I think it makes sense to do it. Could you apply my suggestion and squash everything?

FroMage commented:
I've no idea what this will do. What's the JPA meta-model, and why do we need to generate it?

Postremus commented:
@FroMage See Example 1 and Example 2 from https://docs.jboss.org/hibernate/orm/5.0/topical/html/metamodelgen/MetamodelGenerator.html.

The metamodel allows you to simplify your criteriaquery.

For example, I can write:
`where.add(builder.not(historyEntry.get(HistoryEntry_.eventType).in(filter.getExlusiveEventTypes())));`

Instead of:
`where.add(builder.not(historyEntry.get("eventType").in(filter.getExlusiveEventTypes())));`.

This saves me from referencing "entityType" as String everywhere.
It also allows type safety.

FroMage commented:
OK, thanks. I guess it can't hurt, but I'm not familiar with this at all, so asking @Sanne and @emmanuelbernard for advice.

gsmet commented:
Yeah, it's useful if you use JPA criterias. If you have nothing against it, I will merge it once the typo is fixed.

Sanne commented:
Nice!

Postremus committed:
Provide jpa metamodel for PanacheEntity class
When creating the meta model for an entity extending PanacheEntity, no meta model PanacheEntity_ is found.

Postremus commented:
@gsmet done

gsmet commented:
Merged, thanks!

geoand commented:
I think it would be nice to have a section in the Panache guide showing how this could be used? I personally don't know much about the meta-model and would like to know how to use it :)

emmanuelbernard commented:
Nan, maybe in the normal Hibernate doc. This tool should be in the
Hibernate ORM extension anyways, not in the Panache one as it's generally
useful.

On Tue, Feb 18, 2020 at 3:01 PM Georgios Andrianakis <
notifications@github.com> wrote:

> I think it would be nice to have a section in the Panache guide showing
> how this could be used? I personally don't know much about the meta-model
> and would like to know how to use it :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/7218?email_source=notifications&email_token=AACJNWHDUWVG7A3CGJYFAZDRDPS27A5CNFSM4KWFOUZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMCCHYQ#issuecomment-587473890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACJNWBYV5PC427HDORHQIDRDPS27ANCNFSM4KWFOUZA>
> .
>

gsmet commented:
The idea here is to provide the meta model for `PanacheEntity`.

geoand commented:
> The idea here is to provide the meta model for `PanacheEntity`.

Exactly. So perhaps some quick info on how to use it with Panache and then perhaps a link to the extensive Hibernate documentation?
